src/uadk_digest.c

```diff
@@ -927,3 +1017,153 @@ void uadk_e_destroy_digest(void)
+{
+	struct digest_priv_ctx *ctx;
+
+	ctx = (struct digest_priv_ctx *)dctx;
+	OPENSSL_clear_free(ctx, sizeof(*ctx));
+}
+
+static void *uadk_prov_dupctx(void *dctx)
+{
+	struct digest_priv_ctx *in;
+	struct digest_priv_ctx *ret;
+
+	in = (struct digest_priv_ctx *)dctx;
+	ret = OPENSSL_zalloc(sizeof(struct digest_priv_ctx *));
+
+	if (ret != NULL)
+		*ret = *in;
+	return ret;
+}
+
+static int uadk_prov_init(void *dctx, const OSSL_PARAM params[])
+{
+	struct digest_priv_ctx *priv =
+		(struct digest_priv_ctx *) dctx;
+
+	return uadk_prov_digest_init(priv);
+}
+
+static int uadk_prov_update(void *dctx, const unsigned char *in, size_t inl)
+{
+	return uadk_prov_digest_update((struct digest_priv_ctx *)dctx,
+				       in, inl);
+}
+
+/*
+ * Note:
+ * The I<dctx> parameter contains a pointer to the provider side context.
+ * The digest should be written to I<*out> and the length of the digest to I<*outl>.
+ * The digest should not exceed I<outsz> bytes.
+ */
+static int uadk_prov_final(void *dctx, unsigned char *out, size_t *outl,
+				  size_t outsz)
+{
+	struct digest_priv_ctx *priv =
+		(struct digest_priv_ctx *) dctx;
+
+	if (outsz > 0)
+		uadk_prov_digest_final(priv, out);
+
+	return priv->md_size;
+}
+
+/* Forward declaration of uadk implementation functions */
+static OSSL_FUNC_digest_freectx_fn	uadk_prov_freectx;
+static OSSL_FUNC_digest_dupctx_fn	uadk_prov_dupctx;
+static OSSL_FUNC_digest_init_fn		uadk_prov_init;
+static OSSL_FUNC_digest_update_fn	uadk_prov_update;
+static OSSL_FUNC_digest_final_fn	uadk_prov_final;
+static OSSL_FUNC_digest_gettable_params_fn
+					uadk_prov_gettable_params;
+
+#define UADK_PROVIDER_IMPLEMENTATION(name, nid, mdsize, blksize)		\
+static OSSL_FUNC_digest_newctx_fn uadk_##name##_newctx;				\
+static void *uadk_##name##_newctx(void *provctx)				\
+{										\
+	struct digest_priv_ctx *ctx = OPENSSL_zalloc(sizeof(*ctx));		\
+										\
+	if (ctx == NULL)							\
+		return NULL;							\
+	ctx->blk_size = blksize;						\
+	ctx->md_size = mdsize;							\
+	ctx->e_nid = nid;							\
+	return ctx;								\
+}										\
+static OSSL_FUNC_digest_get_params_fn uadk_##name##_get_params;			\
+static int uadk_##name##_get_params(OSSL_PARAM params[])			\
+{										\
+	return uadk_digest_default_get_params(params, blksize, mdsize);		\
+}										\
+const OSSL_DISPATCH uadk_##name##_functions[] = {				\
+	{ OSSL_FUNC_DIGEST_NEWCTX, (void (*)(void))uadk_##name##_newctx },	\
+	{ OSSL_FUNC_DIGEST_FREECTX, (void (*)(void))uadk_prov_freectx },	\
+	{ OSSL_FUNC_DIGEST_DUPCTX, (void (*)(void))uadk_prov_dupctx },		\
+	{ OSSL_FUNC_DIGEST_INIT, (void (*)(void))uadk_prov_init },		\
+	{ OSSL_FUNC_DIGEST_UPDATE, (void (*)(void))uadk_prov_update },		\
+	{ OSSL_FUNC_DIGEST_FINAL, (void (*)(void))uadk_prov_final },		\
+	{ OSSL_FUNC_DIGEST_GET_PARAMS,						\
+		(void (*)(void))uadk_##name##_get_params },			\
+	{ OSSL_FUNC_DIGEST_GETTABLE_PARAMS,					\
+		(void (*)(void))uadk_digest_gettable_params },			\
```

Comment:
The name "uadk_digest_gettable_params" maybe "uadk_prov_digest_gettable_params".